message.go

```diff
@@ -158,3 +158,31 @@ func (messages *messages) append(item *Message) {
 
 	messages.items = append(messages.items, item)
 }
+
+// Returns a copy of all messages
+func (messages *messages) copy() []Message {
+	messages.RLock()
+	defer messages.RUnlock()
+
+	copiedMessages := []Message{}
+	for index := range messages.items {
+		copiedMessages = append(copiedMessages, *messages.items[index])
+	}
+
+	return copiedMessages
+}
+
+// Returns all messages and removes them at the same time
+func (messages *messages) purge() []Message {
+	messages.Lock()
+	defer messages.Unlock()
+
+	copiedMessages := []Message{}
```

Comment:
We can DRY this part of code:

```go
func (messages *messages) copyMessages() []Message {
	copiedMessages := []Message{}
	for index := range messages.items {
		copiedMessages = append(copiedMessages, *messages.items[index])
	}

	return copiedMessages
}


func (messages *messages) copy() []Message {
	messages.RLock()
	defer messages.RUnlock()
	copiedMessages := messages.copyMessages()

	return copiedMessages
}

func (messages *messages) purge() []Message {
	messages.Lock()
	defer messages.Unlock()
	copiedMessages := messages.copyMessages()
	messages.items = nil

	return copiedMessages
}
```

Reply:
Sadly that will not work well. Once you acquire exlusive lock in your `purge`, `copyMessages` will block on `messages.RLock()`.

Reply:
We could move the main copying of `copyMessages` into another helper without any locking.

Reply:
> We could move the main copying of `copyMessages` into another helper without any locking, but it is short so not sure how much we gain by doing that.

Ah, my bad. Good catch! Yes the idea is to move to another helper. Please skip it for now.